File: depth_first_search.java
import java.util.ArrayList;

public class depth_first_search {

    static class Edge{
        int source;
        int destination;

        Edge(int s, int d){
            this.source=s;
            this.destination=d;
        }
    }
    public static void creategraph(ArrayList<Edge> graph[]){
        for(int i=0;i<graph.length;i++){
            graph[i]=new ArrayList<>();
        }
        graph[0].add(new Edge(0,1));
        graph[0].add(new Edge(0,2));
        graph[1].add(new Edge(1,3));
        graph[2].add(new Edge(2,4));
        graph[3].add(new Edge(3,4));
        graph[3].add(new Edge(3,5));
        graph[3].add(new Edge(3,2));
        graph[4].add(new Edge(4,2));
        graph[4].add(new Edge(4,5));
        graph[5].add(new Edge(5,6));
    }

    //DFS function
    public static void dfs(ArrayList<Edge> graph[],boolean[]visited, int current){
        System.out.print(current + " ");
        visited[current]=true;

        for(int i=0;i<graph[current].size();i++){
            Edge e=graph[current].get(i);
            if(!visited[e.destination]){
                dfs(graph,visited,e.destination);
            }
        }
    }
    public static void main(String[] args) {

        int vertices = 7;
        ArrayList<Edge>[] graph = new ArrayList[vertices];
        creategraph(graph);

        boolean[] visited = new boolean[vertices];
        System.out.println("Depth First Search starting from vertex 0:");
        dfs(graph, visited, 0);
    }
}

File: depth_first_search.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

void addEdge(vector<int> adj[], int source,int destination)
{
    adj[source].push_back(destination);
    adj[destination].push_back(source);
}

void AdjacencyList(vector<int> adj[],int vertices)
{
    for(int i=0;i<vertices;i++)
    {
        cout<<i;
        for(int j=0;j<adj[i].size();j++)
        {
            cout<<"->"<<adj[i][j];
        }
        cout<<"\n";
    }
}



//Iterative approach
void DFS(vector<int>adj[],int start,int vertices)
{
    vector<bool>visited(vertices,false);
    stack<int>s;

    s.push(start);
    while(!s.empty())
    {
        int current=s.top();
        s.pop();
        if(!visited[current])
        {
            visited[current]=true;
            cout<<current<<" ";
        }
        for(int i=0;i<adj[current].size();i++)
        {
            int neighbor=adj[current][i];
            if(!visited[neighbor])
            {
                s.push(neighbor);
            }
        }
    }
}
int main() {
  int V = 11;

  vector<int> adj[V];
  addEdge(adj, 0, 1);
  addEdge(adj, 0, 3);
  addEdge(adj, 0, 6);
  addEdge(adj, 1, 2);
  addEdge(adj, 1, 4);
  addEdge(adj, 1, 4);
  addEdge(adj, 1, 4);
  addEdge(adj, 4, 5);

  AdjacencyList(adj, V);
  DFS(adj,0,V);

  return 0;
}
